Fixed returns doc for init methods issue-557

mohpor opened the pull request:
Related to #557 
There might be a better solution, but this pull request does the job, simple and easy.

freak4pc commented on Source/SwiftLintFramework/Rules/ValidDocsRule.swift:
```diff
@@ -51,7 +51,9 @@ func declarationReturns(declaration: String, kind: SwiftDeclarationKind? = nil)
         return false
     }
-
+    if declaration.containsString("init(") {
```
This is dangerous, Consider a declaration of sorts:
`actionPerformFinit()` 
This would still match this case incorrectly. 

I think something like this would be better for this "Simple" style solution but I'm still not totally sure if checking for a string match is the best solution regardless. I think it could work for now if @jpsim thinks this is proper. 

```swift
guard !declaration.hasPrefix("init(") else {
    return true
}
```

mohpor replied:
@freak4pc You are right, should've  considered the constant string more carefully.
* I'm not that familiar with the project to be honest, and I wasn't aware that the `declaration`  string would be something like: `public convince init()` or plain `init()`, that's why I didn't go for the `hasPrefix`, but I see your point and I think a leading space could help!
* I decided against the `guard` usage instead of `if` because the guidelines I'm following, prohibits using guards for case testing and this is clearly testing against different scenarios, but, if SwiftLint suggests I should use `guard`, by all means!
* The very next line of code uses the same technique for checking if a function returns (by checking the existence of `->`), granted, that is very much unique in the swift grammar, but then, so is `init()`, don't you agree?

freak4pc replied:
I think the beginning space solution should be good :) 

I'm also not familiar enough with SourceKitten so im not sure if includes the ACL etc but it should be considered in my opinion.  

About guard/if, As far as I know, [the guidelines SwiftLint follows](https://github.com/github/swift-style-guide/blob/master/README.md#return-and-break-early), do suggest using guard to return early.

mohpor replied:
Thanks for the guideline suggestion. That's very helpful. Wilco. :)

jpsim commented:
This wouldn't work for failable initializers (`init?`). You should check if the declaration is a [`SwiftDeclarationKind.FunctionConstructor`](https://github.com/jpsim/SourceKitten/blob/0.13.0/Source/SourceKittenFramework/SwiftDeclarationKind.swift#L45) instead.

Grubas7 commented:
@jpsim Agree, I tried this solution but `kind` is never equal to `.FunctionConstructor`.
For code:
```
class TestClass: NSObject {
    
    /**
     Description about TestClass
     - parameter string: a super string
     - returns: a new class
     */
    override init() {
        super.init()
    }
}
```

returns `.FunctionMethodInstance`. I don't know why whole `dictionary` doesn't contain `. FunctionConstructor` value.

jpsim commented:
Ah, interesting. That may be resolved later in the lexer than what's available to SourceKit.

@mohpor could you at least update this to support both `init` and `init?` please?

mohpor commented:
@jpsim Thanks for the heads up.
I have updated the checking mechanism to support failable initializers as well, made it a separate function, and moved checking to bail out of checking earlier.

The point here IMHO is that, a `- returns:` documentation for constructors should be **optional**. Because a constructor has a very specific job. To force the existence of returns doc makes documentation verbose. That being said, there are cases where you want to explain something about the constructors job (especially in case of failable ones), that should be possible too, so I guess a returns doc for constructors should be optional, not mandatory, nor prohibited.

P.S.: I tried using regex to match the initializer syntax but I failed miserably!
Here is the regex code I was trying to use:

```swift
func declarationIsContructor(declaration: String) -> Bool {
    return regex("\\sinit\\?*\\(.*\\)").numberOfMatchesInString(declaration, options: [], range: NSRange(location: 0, length: declaration.characters.count)) > 0
}
```

But I couldn't make it work! It wouldn't match ` init()` or ` init?()` no matter what!

freak4pc commented:
Just a heads up that this could potentially cause an issue with any method name that ends with the letters `init` ... not just initializers. So when the SourceKitten issue is resolved we should definitely switch to that. 

About the regex, this should work fine `(.*)(init|init?)`, here's a short Playground test I did.

![image](https://cloud.githubusercontent.com/assets/605076/17356066/d4bf86dc-595d-11e6-9861-979c0a92c0ef.png)

freak4pc commented:
For the regex thing, this should work if you're interested in doing it
`(.*)init(\\?)?\\((.*)\\)` 

Regex in Swift is painful though. 

Here's a short Playground test: 
![image](https://cloud.githubusercontent.com/assets/605076/17356209/c96242e2-595e-11e6-8b74-892cdd17f85e.png)

freak4pc commented:
About your whitespace issue, 
`Xcode -> Preferences -> Text Editing -> Automatically trim trailing whitespace, including whitespace-only lines` 

;-)

mohpor commented:
The regex you've mentioned would match `performinit()` too. The leading space should be mandatory I think. a better approach would be to start from the beginning and move from there, check if the is any character and is not a whitespace, then check that before `init` there is a whitespace. That makes for a relatively complex regex.

P.S.: Strangely enough, the regex I have provided before works in playground but fails in action! I will take a closer look at this.

P.P.S: thanks a lot for guiding me through that vicious trailing white space, it's been bugging me for too long. 👍

mohpor commented:
The correct regex I think is something like this: `^((.+)?\s+)?init\?*\(.*\)` do you find any caveats whit this @freak4pc ?

freak4pc commented:
Nothing that I can notice, it looks good here with `^((.+)?\\s+)?init\\?*\\(.*\\)` 

![image](https://cloud.githubusercontent.com/assets/605076/17356496/706a4782-5960-11e6-9672-a4b6c6bd1125.png)

mohpor commented:
Worked like a charm!

mohpor commented:
@jpsim do you need anything to add to this one?

mohpor commented:
It's been 8 days! Any news?

Grubas7 commented:
@mohpor Could you make `- returns:` for constructors should be optional, update CHANGELOG.md and squash your commits?

mohpor commented:
@Grubas7 I have updated the CHANGELOG.md.
I thought my commits would be squashed by the admin who will merge this pull request. Am I wrong?

Grubas7 commented:
@mohpor Yeap, you're right. I forgot about this option on GitHub :)

jpsim commented on CHANGELOG.md:
```diff
@@ -20,11 +20,15 @@
   [bootstraponline](https://github.com/bootstraponline)
   [#689](https://github.com/realm/SwiftLint/issues/689)
 
+* Made `- returns:` doc optional for constructors.
```
missing two trailing spaces as indicated in https://github.com/realm/SwiftLint/blob/master/CONTRIBUTING.md#tracking-changes

I'd also prefer if you replaced the word "constructors" with "initializers" because constructors also include the family of static functions that return `Self` and this doesn't apply to that.

jpsim commented on CHANGELOG.md:
```diff
@@ -49,7 +53,7 @@
 
 * Remove extraneous argument label added in LegacyCGGeometryFunctionsRule
-  autocorrect. 
+  autocorrect.
```
there should also be two trailing spaces here

jpsim commented on Source/SwiftLintFramework/Rules/ValidDocsRule.swift:
```diff
@@ -66,6 +65,12 @@ func matchOutsideBraces(declaration: String) -> NSString? {
     return NSString(string: declaration).substringWithRange(outsideBracesMatch.range)
 }
 
+func declarationIsConstructor(declaration: String) -> Bool {
```
`declarationIsInitializer`?

jpsim commented:
@mohpor thanks for your work on this, and sorry for the delay on reviewing this. I just had a few minor comments. Could you please address them, then I think this can be merged.

mohpor commented:
@jpsim Done.

P.S.: My branch has conflicts now. Should I do anything?

freak4pc commented:
@mohpor you'll probably have to rebase against master

mohpor commented:
@freak4pc Thanks again. You're the man.

freak4pc commented:
I think you made a booboo @mohpor :P

mohpor commented:
What?! Where?! This is one messy PR!
:-(![](https://link.nylas.com/open/afgeqdx9k5tlkg3470i2id8pp/local-a9a8ae3c-f7
12?r=cmVwbHkrMDA0Yjk5ZGMzNGE1ZWIyMzg3MjI0MDMwY2NhMjE0Y2JiZDgxYmZjMTM3ZjE3NTczO
TJjZjAwMDAwMDAxMTNkMmFjYzg5MmExNjljZTBhMGZjZWIyQHJlcGx5LmdpdGh1Yi5jb20=)

  
On Aug 22 2016, at 4:36 pm, Shai Mishali <notifications@github.com> wrote:  

> I think you made a booboo [@mohpor](https://github.com/mohpor&r=cmVwbHkrMDA0
Yjk5ZGMzNGE1ZWIyMzg3MjI0MDMwY2NhMjE0Y2JiZDgxYmZjMTM3ZjE3NTczOTJjZjAwMDAwMDAxMT
NkMmFjYzg5MmExNjljZTBhMGZjZWIyQHJlcGx5LmdpdGh1Yi5jb20=) :P

>

> --  
You are receiving this because you were mentioned.  
Reply to this email directly, [view it on GitHub](https://github.com/realm/Swi
ftLint/pull/739#issuecomment-241393092&r=cmVwbHkrMDA0Yjk5ZGMzNGE1ZWIyMzg3MjI0M
DMwY2NhMjE0Y2JiZDgxYmZjMTM3ZjE3NTczOTJjZjAwMDAwMDAxMTNkMmFjYzg5MmExNjljZTBhMGZ
jZWIyQHJlcGx5LmdpdGh1Yi5jb20=), or [mute the
thread](https://github.com/notifications/unsubscribe-auth/AEuZ3DBsHFCeTjGE_xs_
uB_5hwigg5j1ks5qiZDIgaJpZM4JaR22&r=cmVwbHkrMDA0Yjk5ZGMzNGE1ZWIyMzg3MjI0MDMwY2N
hMjE0Y2JiZDgxYmZjMTM3ZjE3NTczOTJjZjAwMDAwMDAxMTNkMmFjYzg5MmExNjljZTBhMGZjZWIyQ
HJlcGx5LmdpdGh1Yi5jb20=).![](https://github.com/notifications/beacon
/AEuZ3MNzJ2_VT-OV6JXTIiWUqHuQqi_kks5qiZDIgaJpZM4JaR22.gif)

mohpor commented:
Hmmmm, rebasing made everyone's commits into this one. I guess I need to close this commit, make a clean, one commit PR instead. Do you agree @freak4pc @jpsim ?

jpsim commented:
> Hmmmm, rebasing made everyone's commits into this one. I guess I need to close this commit, make a clean, one commit PR instead.

You don't need to close this PR, you can rebase (correctly) and force-push to the same branch.

mohpor commented:
@jpsim Sorry for the mixup, I have re-rebased and squashed my commits.
It seems like a simple merge but I learned a ton.
I should thank @freak4pc for his advices too, very helpful.

freak4pc commented:
Anytime. Thanks for your consistent help on this PR! @mohpor

jpsim commented:
🎉 congrats for landing this @mohpor

Grubas7 commented:
This bug is still valid for implicitly unwrapped optional, eg:
`init!()`

jpsim commented:
huh, forgot it was possible to make initializers like that.

mohpor commented:
Whoops,
Will make a PR soon.

mohpor commented:
Checkout #804

norio-nomura commented:
We can make initializers generic.
